sys/kern/vm_map.c

```diff
@@ -371,15 +372,42 @@ vm_map_t *vm_map_clone(vm_map_t *map) {
     TAILQ_FOREACH (it, &map->entries, link) {
       vm_object_t *obj;
       vm_segment_t *seg;
+      vm_object_t *backing;
+
+      SCOPED_MTX_LOCK(&it->object->mtx);
 
       if (it->flags & VM_SEG_SHARED) {
         refcnt_acquire(&it->object->ref_counter);
         obj = it->object;
       } else {
-        /* vm_object_clone will clone the data from the vm_object_t
-         * and will return the new object with ref_counter equal to one */
-        obj = vm_object_clone(it->object);
+        /* check if this object didn't copy any page, if so,
+         * then we don't have to create another element in vm_objects chain,
+         * we can just use existing one */
+        if ((it->flags & VM_SEG_COW) != 0 && it->object->npages == 0) {
+          obj = vm_object_alloc(VM_SHADOW);
+          backing = it->object->backing_object;
+          obj->backing_object = backing;
+        } else {
+          backing = it->object;
+          obj = vm_object_alloc(VM_SHADOW);
+          obj->backing_object = backing;
+          it->object = vm_object_alloc(VM_SHADOW);
+          it->object->backing_object = backing;
+          /* we set a backing object twice, so we have to acquire
+           * backing->shadow_counter twice too */
+          refcnt_acquire(&backing->shadow_counter);
+          /* all pages in backing object are now read-only,
+           * that refers also to pages which previously had VM_PROT_EXEC set
+           */
+          vm_object_set_prot(backing, VM_PROT_READ);
```

Comment:
I'm concerned about how this would interact with pages mapped without `VM_PROT_READ`... For instance pages mapped as execute-only (I don't know if it's a thing on `aarch64`), or guard pages with `VM_PROT_NONE`: when cloning such a segment, you would enable read access in the page table, and read access wouldn't cause a page fault.

Reply:
Okay, I added checking if segment this object originally belonged to, has `VM_PROT_READ` set and if so, then I set all pages as readonly.